File: 1475 final-prices-with-a-special-discount-in-a-shop/solution.java
class Solution {
    public int[] finalPrices(int[] prices) {
        int[] st = new int[prices.length];
        int top = 0,idx=0;
        int[] ans = new int[prices.length];
        for(int i=1;i<prices.length;i++) {
            while(top!=-1 && prices[st[top]] >= prices[i]) {
                ans[st[top]] = prices[st[top--]] - prices[i]; 
            }
            st[++top] = i;
        }
        for(int i=0;i<=top;i++) ans[st[i]] = prices[st[i]];
        return ans;
    }
}

File: 1475 final-prices-with-a-special-discount-in-a-shop/solution.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        int n = prices.size();
        vector<int> ans(n);
        vector<int> st; 
        for (int i = 0; i < n; i++) {
            while (!st.empty() && prices[st.back()] >= prices[i]) {
                int idx = st.back();
                st.pop_back();
                ans[idx] = prices[idx] - prices[i];
            }
            st.push_back(i);
        }
        for (int idx : st) {
            ans[idx] = prices[idx];
        }
        return ans;
    }
};

